drivers/base/nmidebug/nmidebug.c

```diff
@@ -13,49 +12,31 @@
 
 /* FUNCTIONS ******************************************************************/
 
-PCHAR NmiBegin = "NMI4NMI@";
-
-FORCEINLINE
-VOID
-NmiClearFlag(VOID)
-{
-    ((PCHAR)&KiBugCheckData[4])[0] -= (NmiBegin[3] | NmiBegin[7]);
-    ((PCHAR)&KiBugCheckData[4])[3] |= 1;
-#ifdef _M_IX86
-#if defined(_MSC_VER) && !defined(__clang__)
-    __asm
-    {
-        rcr KiBugCheckData[4], 8
-    }
-#else
-    __asm__("rcrl %b[shift], %k[retval]" : [retval] "=rm" (KiBugCheckData[4]) : "[retval]" (KiBugCheckData[4]), [shift] "Nc" (8));
-#endif
-#endif
-}
-
 BOOLEAN
 NTAPI
-NmiDbgCallback(IN PVOID Context,
-               IN BOOLEAN Handled)
+NmiDbgCallback(
+    _In_opt_ PVOID Context,
+    _In_ BOOLEAN Handled)
 {
-    /* Clear the NMI flag */
-    NmiClearFlag();
+    /* Handle the NMI safely if a debugger is connected */
+    if (!KD_DEBUGGER_NOT_PRESENT && KD_DEBUGGER_ENABLED)
+    {
+        DbgPrint("NMI Callback entered!\n");
 
-    /* Get NMI status signature */
-    __indwordstring(0x80, (PULONG)NmiBegin, 1);
-    ((void(*)())&KiBugCheckData[4])();
+        DbgBreakPoint();
+
+        return TRUE;
+    }
 
-    /* Handle the NMI safely */
-#ifdef _M_IX86
-    KiEnableTimerWatchdog = (RtlCompareMemory(NmiBegin, NmiBegin + 4, 4) != 4);
```

Comment:
Somehow, this checks whether the modified `"NMI4"` part of NmiBegin is equal to the second (untouched) `"NMI@"` part.